### PR TITLE
test: always perform assertions in module API test

### LINKED ISSUE #7625: module_api.c unused parameter/variable warnings
The problem is introduced in 5c1bc3dacbe6814e877ca48d50c094ec79453773. `2.10` and `master` branches are affected.

```c
/path/to/tarantool/test/app-tap/module_api.c: In function ‘check_decimal’:
/path/to/tarantool/test/app-tap/module_api.c:2437:53: warning: unused parameter ‘exp’ [-Wunused-parameter]
 2437 | check_decimal(const box_decimal_t *dec, const char *exp)
      |                                         ~~~~~~~~~~~~^~~
/path/to/tarantool/test/app-tap/module_api.c: In function ‘poison_check’:
/path/to/tarantool/test/app-tap/module_api.c:2482:21: warning: unused variable ‘poison_after’ [-Wunused-variable]
 2482 |         const char *poison_after = raw + sizeof(size_t) + POISON_SIZE + size;
      |                     ^~~~~~~~~~~~
/path/to/tarantool/test/app-tap/module_api.c:2481:21: warning: unused variable ‘poison_before’ [-Wunused-variable]
 2481 |         const char *poison_before = raw + sizeof(size_t);
      |                     ^~~~~~~~~~~~~
/path/to/tarantool/test/app-tap/module_api.c: In function ‘test_decimal’:
/path/to/tarantool/test/app-tap/module_api.c:2850:15: warning: unused variable ‘data_end’ [-Wunused-variable]
 2850 |         char *data_end = box_decimal_mp_encode(&ammonia, data);
      |               ^~~~~~~~
/path/to/tarantool/test/app-tap/module_api.c:2608:13: warning: variable ‘rc’ set but not used [-Wunused-but-set-variable]
 2608 |         int rc = box_decimal_compare(&five_1, &six);
      |             ^~
/path/to/tarantool/test/app-tap/module_api.c:2589:30: warning: variable ‘cp’ set but not used [-Wunused-but-set-variable]
 2589 |         const box_decimal_t *cp = box_decimal_to_int64(&carthage, &carthage_64);
      |                              ^~
/path/to/tarantool/test/app-tap/module_api.c:2558:14: warning: variable ‘is_neg’ set but not used [-Wunused-but-set-variable]
 2558 |         bool is_neg = box_decimal_is_neg(&pi);
      |              ^~~~~~
/path/to/tarantool/test/app-tap/module_api.c:2548:14: warning: variable ‘is_int’ set but not used [-Wunused-but-set-variable]
 2548 |         bool is_int = box_decimal_is_int(&pi);
      |              ^~~~~~
/path/to/tarantool/test/app-tap/module_api.c:2538:13: warning: unused variable ‘scale’ [-Wunused-variable]
 2538 |         int scale = box_decimal_scale(&pi);
      |             ^~~~~
/path/to/tarantool/test/app-tap/module_api.c:2527:13: warning: variable ‘precision’ set but not used [-Wunused-but-set-variable]
 2527 |         int precision = box_decimal_precision(&pi);
      |             ^~~~~~~~~
/path/to/tarantool/test/app-tap/module_api.c:2511:24: warning: variable ‘p’ set but not used [-Wunused-but-set-variable]
 2511 |         box_decimal_t *p = box_decimal_from_string(&pi, "3.14");
```

### PR DESCRIPTION
**Pull request by Totktonada**
This commit pursues several goals:

* Eliminate unused parameter/variable warnings at building module_api.c
  in non-debug configuration. The problem was introduced in commit
  5c1bc3dacbe6 ("decimal: add the library into the module API").
* Eliminate a need to check newly added tests in two build
  configurations (Debug and RelWithDebInfo) and don't forget to add
  `(void)x;` statements in addition to a test condition check.
* Fail the testing if conditions required by the
  app-tap/module_api.test.lua test are not met -- not only in the Debug
  build, but also in RelWithDebInfo.

Fixes #7625

### CODE REVIEW

**Comment by Totktonada** [comment]
@locker, @sergepetrenko, I assigned you as reviewers, because you recently looked at the patch, where I made the mistake that is fixed there.

There is no any urgency here. Please, look at spare time or unassign youself from reviewers if the time is tight.

File: test/app-tap/module_api.c [hunk 1]
```diff
@@ -1,3 +1,10 @@
+/*
+ * Use assert() as a testing assertion function disregarding
+ * tarantool's build type.
+ */
+#undef NDEBUG
```
Comment (locker) [inline test/app-tap/module_api.c:5]: I'm not against it, but this looks weird. Maybe, replace the assertions with `fail_if` from `unit.h` instead?

BTW personally I'm quite used to void-ing asserted variables (we do it all the time in the Tarantool source code). Lack of this `void` casts looks uneasy to me.

Comment (Totktonada) [reply test/app-tap/module_api.c:5]: No problem, I'll add `fail_if` macro into the file and will replace `assert()`s with it.

Comment (Totktonada) [reply test/app-tap/module_api.c]: Replaced `assert()` calls with `fail_unless()`.

**Review by locker — APPROVED** [review]

**Review by sergepetrenko — APPROVED** [review]

**Comment by locker** [comment]
Cherry-picked to 2.10.

**Pull request merged**